src/loading.test.ts

```diff
@@ -0,0 +1,87 @@
+import { TestScheduler } from 'rxjs/testing'
+import { emitLoading, LOADING, MaybeLoadingResult } from './loading'
+
+const deepStrictEqual = chai.assert.deepStrictEqual.bind(chai.assert)
+
+const inputAlphabet: Record<'l' | 'e' | 'i' | 'r', MaybeLoadingResult<number | null>> = {
+    // loading
+    l: { isLoading: true, result: null },
+    // empty
+    e: { isLoading: false, result: null },
+    // intermediate result
+    i: { isLoading: true, result: 1 },
+    // result
+    r: { isLoading: false, result: 2 },
+}
+
+const outputAlphabet = {
+    // undefined
+    u: undefined,
+    // empty
+    e: null,
+    // loading
+    l: LOADING,
+    // intermediate result
+    i: inputAlphabet.i.result,
+    // result
+    r: inputAlphabet.r.result,
+}
+
+describe('emitLoading()', () => {
+    it('emits an empty result if the source emits an empty result before the loader delay', () => {
+        const scheduler = new TestScheduler(deepStrictEqual)
+        scheduler.run(({ cold, expectObservable }) => {
+            const source = cold('l 10ms e', inputAlphabet)
+            expectObservable(source.pipe(emitLoading(300, null))).toBe('u 10ms e', outputAlphabet)
+        })
+    })
+    it('emits a loader if the source has not emitted after the loader delay', () => {
+        const scheduler = new TestScheduler(deepStrictEqual)
+        scheduler.run(({ cold, expectObservable }) => {
+            const source = cold('400ms r', inputAlphabet)
+            expectObservable(source.pipe(emitLoading(300, null))).toBe('u 299ms l 99ms r', outputAlphabet)
```

Comment:
Out of curiosity: what is the reason these are off by 1ms and other tests by other values? 🤔

Reply:
Good question :) The answer is in the rxjs docs:

> NOTE: You may have to subtract 1 millisecond from the time you want to progress because the alphanumeric marbles (representing an actual emitted value) advance time 1 virtual frame themselves already, after they emit. This can be very unintuitive and frustrating, but for now it is indeed correct.

https://rxjs-dev.firebaseapp.com/guide/testing/marble-testing#time-progression-syntax

Reply:
Ohh wow 👀